compiler find confuses clang-ocl with clang

### Steps to reproduce

I have a one-node RHEL 8.x x86_64 system with an AMD GPU and ROCM
installed.  This is the ROCM tar file downloaded from AMD installed
into /opt/rocm-4.3.1, not a spack install.  There is no clang
compiler, but ROCM installs clang-ocl in /usr/bin.

So, there is no clang and `spack compiler find` confuses clang-ocl
with clang and creates an entry in `compilers.yaml`.

```
compilers:
- compiler:
    spec: clang@13.0.0
    paths:
      cc: /usr/bin/clang-ocl
      cxx: null
      f77: null
      fc: null
    flags: {}
    operating_system: rhel8
    target: x86_64
    modules: []
    environment: {}
    extra_rpaths: []
```

There are two problems here.

1. `clang-ocl` (opencl compiler) is not a regular compiler like gcc,
icc, xl, etc, and spack should not create an entry in compilers.yaml
for it.

For example, `clang-ocl hello.c` on the "hello world" program dies with:

```
$ clang-ocl hello.c 
In file included from hello.c:5:
In file included from /usr/include/stdio.h:27:
In file included from /usr/include/bits/libc-header-start.h:33:
In file included from /usr/include/features.h:452:
/usr/include/gnu/stubs.h:7:11: fatal error: 'gnu/stubs-32.h' file not found
# include <gnu/stubs-32.h>
1 error generated.
```

2. Running `clang-ocl --version` creates a file `-.o`.  I guess it
tries to compile stdin or something, even though you only ask for the
version.

Spack should either not run clang-ocl, or else cleanup the `-.o` file.

--------

I looked at `clang.py` and there is no mention of clang-ocl, but it
does use regular expressions, I guess to look for `clang-*`.  On
systems with clang-ocl but not clang, this causes spack to confuse
clang-ocl with the real (and absent) clang compiler.

Basically, spack should understand that clang-ocl is not a valid
compiler that it's looking for.

But looking in `/opt/rocm-4.3.1/llvm/bin`, there are other valid
compiler names that are symlinked to clang or clang-13, including
flang, just not clang-ocl.

```
lrwxrwxrwx. 1 root root         8 Aug 21 12:30 clang -> clang-13*
lrwxrwxrwx. 1 root root         5 Aug 21 12:30 clang++ -> clang*
-rwxr-xr-x. 1 root root 114674512 Aug 21 12:30 clang-13*
lrwxrwxrwx. 1 root root         5 Aug 21 12:30 clang-cl -> clang*
lrwxrwxrwx. 1 root root         5 Aug 21 12:30 clang-cpp -> clang*
lrwxrwxrwx. 1 root root         5 Aug 21 12:30 flang -> clang*
```

Also, I'm seeing `spack compiler find` behave non-deterministically.
Sometimes it finds one compiler (either one) and sometimes two.



### Error message

[compiler-find.txt](https://github.com/spack/spack/files/7138843/compiler-find.txt)


### Information on your system

* **Spack:** 0.16.2-4197-f9314d38b0
* **Python:** 3.6.8
* **Platform:** linux-rhel8-zen2
* **Concretizer:** clingo


### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

--- a/var/spack/repos/builtin/packages/llvm/package.py
+++ b/var/spack/repos/builtin/packages/llvm/package.py
@@ -28,7 +28,7 @@ class LlvmDetection(PackageBase):
         # Executables like lldb-vscode-X are daemon listening on some port and would hang Spack
         # during detection. clang-cl, clang-cpp, etc. are dev tools that we don't need to test
         reject = re.compile(
-            r"-(vscode|cpp|cl|gpu|tidy|rename|scan-deps|format|refactor|offload|"
+            r"-(vscode|cpp|cl|ocl|gpu|tidy|rename|scan-deps|format|refactor|offload|"
             r"check|query|doc|move|extdef|apply|reorder|change-namespace|"
             r"include-fixer|import-test|dap|server)"
         )

